File: apps/browser-demo/src/deviceAgent/__tests__/geminiNativeProvider.test.ts
// Pins the Gemini native :generateContent wire shape for the new Device Agent path:
// URL is `${baseUrl}/models/${model}:generateContent` (with `/openai` stripped from
// the preset baseUrl), x-goog-api-key header (not Authorization), systemInstruction
// + contents[].parts[] body, responseMimeType: 'application/json' on plan/review only,
// and google_search tool attached only on the research pass (with responseMimeType
// removed in the same condition — Gemini rejects the combination). The Kotlin port
// for this class is a followup ticket.

import { describe, expect, it } from 'vitest';

import { buildPlanMessages, buildReviewMessages } from '../prompts/messageAssembler.js';
import { DEVICE_AGENT_SYSTEM_PROMPTS } from '../prompts/systemPrompts.js';
import type { RuntimeConfig } from '../runtime/config.js';

import { ProviderHttpError } from '../provider/errorCodes.js';
import { GeminiNativeProvider } from '../provider/geminiNativeProvider.js';

import { FakeHttpExecutor } from './fakeHttpExecutor.helper.js';

function config(model = 'gemini-2.5-pro', baseUrl = 'https://generativelanguage.googleapis.com/v1beta/openai'): RuntimeConfig {
  return {
    provider: 'gemini',
    apiFormat: 'openai-compatible',
    model,
    baseUrl,
    apiKey: 'AIzaTEST-ABCDEFGHIJKLMNOP',
  };
}

function geminiTextResponse(text: string): string {
  return JSON.stringify({
    candidates: [{ content: { parts: [{ text }] } }],
  });
}

describe('GeminiNativeProvider.generatePlan', () => {
  it('POSTs to :generateContent with systemInstruction, x-goog-api-key, and responseMimeType for plan', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(
      200,
      geminiTextResponse(
        '{"intent":"swap","route":"jupiter","risk":"low","approval":"once","safeguards":["check slippage"]}',
      ),
    );
    const provider = new GeminiNativeProvider(config(), http);
    const result = await provider.generatePlan({ userPrompt: 'swap 1 SOL for USDC' });

    expect(result.intent).toBe('swap');
    const call = http.calls[0]!;
    // /openai suffix stripped from the preset baseUrl; model + :generateContent appended.
    expect(call.url).toBe(
      'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-pro:generateContent',
    );
    expect(call.headers['x-goog-api-key']).toBe('AIzaTEST-ABCDEFGHIJKLMNOP');
    expect(call.headers.Authorization).toBeUndefined();

    const body = JSON.parse(call.body) as Record<string, unknown>;
    expect(body.systemInstruction).toEqual({ parts: [{ text: DEVICE_AGENT_SYSTEM_PROMPTS.PLAN }] });
    const contents = body.contents as Array<{ role: string; parts: Array<{ text: string }> }>;
    expect(contents).toHaveLength(1);
    expect(contents[0]!.role).toBe('user');
    expect(contents[0]!.parts[0]!.text).toBe(buildPlanMessages({ userPrompt: 'swap 1 SOL for USDC' }).userContent);
    const generationConfig = body.generationConfig as Record<string, unknown>;
    expect(generationConfig.temperature).toBe(0.2);
    expect(generationConfig.maxOutputTokens).toBe(1024);
    expect(generationConfig.responseMimeType).toBe('application/json');
    expect('tools' in body).toBe(false);
  });

  it('accepts an already-native baseUrl without an /openai suffix (idempotent)', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, geminiTextResponse('{"intent":"x"}'));
    const provider = new GeminiNativeProvider(
      config('gemini-2.5-pro', 'https://generativelanguage.googleapis.com/v1beta'),
      http,
    );
    await provider.generatePlan({ userPrompt: 'hi' });
    expect(http.calls[0]!.url).toBe(
      'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-pro:generateContent',
    );
  });

  it('appends /v1beta when caller supplies a bare host', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, geminiTextResponse('{"intent":"x"}'));
    const provider = new GeminiNativeProvider(
      config('gemini-2.5-flash', 'https://example-gemini-proxy.dev'),
      http,
    );
    await provider.generatePlan({ userPrompt: 'hi' });
    expect(http.calls[0]!.url).toBe(
      'https://example-gemini-proxy.dev/v1beta/models/gemini-2.5-flash:generateContent',
    );
  });
});

describe('GeminiNativeProvider.reviewPlan two-pass research', () => {
  it('runs research with google_search and NO responseMimeType, then review with responseMimeType and NO tools', async () => {
    const http = new FakeHttpExecutor();
    // Pass 1: grounded research returns prose + groundingMetadata.groundingChunks.
    http.queueResponse(
      200,
      JSON.stringify({
        candidates: [
          {
            content: {
              parts: [{ text: 'Helium Mobile cheapest plan is Air at $15/month.' }],
            },
            groundingMetadata: {
              groundingChunks: [
                { web: { uri: 'https://www.heliummobile.com/plans', title: 'Plans — Helium Mobile' } },
              ],
            },
          },
        ],
      }),
    );
    // Pass 2: structured review returns approve.
    http.queueResponse(
      200,
      geminiTextResponse(
        '{"decision":"approve","reason":"Air plan is $15, under $20.","summary":"Approved","evidence":{"findings":[{"label":"Subscription price","value":"$15/month","tone":"good"}]}}',
      ),
    );

    const provider = new GeminiNativeProvider(config(), http);
    const result = await provider.reviewPlan({
      instruction: 'check helium mobile. lowest monthly plan. if less than $20. approve.',
      plan: { intent: 'swap' },
      research: { needed: true, mode: 'auto_current_facts', currentDate: '2026-05-17T00:00:00.000Z', maxSearches: 3 },
    });

    expect(result.decision).toBe('approve');
    expect(http.calls.length).toBe(2);

    const researchBody = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    expect(researchBody.tools).toEqual([{ google_search: {} }]);
    const researchGenConfig = researchBody.generationConfig as Record<string, unknown>;
    expect('responseMimeType' in researchGenConfig).toBe(false);
    expect(researchGenConfig.maxOutputTokens).toBe(1800);

    const reviewBody = JSON.parse(http.calls[1]!.body) as Record<string, unknown>;
    expect('tools' in reviewBody).toBe(false);
    const reviewGenConfig = reviewBody.generationConfig as Record<string, unknown>;
    expect(reviewGenConfig.responseMimeType).toBe('application/json');
    const reviewContents = reviewBody.contents as Array<{ parts: Array<{ text: string }> }>;
    expect(reviewContents[0]!.parts[0]!.text).toContain('researchEvidence');
    expect(reviewContents[0]!.parts[0]!.text).toContain('heliummobile.com');
  });

  it('falls back to single-pass review when research pass throws (non-fatal)', async () => {
    const http = new FakeHttpExecutor();
    http.queueFailure(new Error('research pass network blip'));
    http.queueResponse(
      200,
      geminiTextResponse(
        '{"decision":"needs_input","reason":"Cannot confirm current price.","summary":"Needs input","evidence":{}}',
      ),
    );
    const provider = new GeminiNativeProvider(config(), http);
    const result = await provider.reviewPlan({
      plan: { intent: 'swap' },
      research: { needed: true, mode: 'auto_current_facts', currentDate: '2026-05-17T00:00:00.000Z', maxSearches: 3 },
    });

    expect(result.decision).toBe('needs_input');
    expect(http.calls.length).toBe(2);
  });

  it('single-pass review (no research) keeps responseMimeType json and omits tools', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(
      200,
      geminiTextResponse('{"decision":"approve","reason":"ok","summary":"go","evidence":{}}'),
    );
    const provider = new GeminiNativeProvider(config(), http);
    // Supply a pre-built research object so the assembler doesn't auto-generate a fresh
    // `currentDate: new Date().toISOString()` between the provider call and the test's
    // own buildReviewMessages comparison — that mismatch is just timing noise.
    const reviewPayload = {
      plan: { intent: 'swap' },
      research: { needed: false, mode: 'not_required', currentDate: '2026-05-17T00:00:00.000Z', maxSearches: 3 },
    };
    const result = await provider.reviewPlan(reviewPayload);

    expect(result.decision).toBe('approve');
    const body = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    expect('tools' in body).toBe(false);
    expect((body.generationConfig as Record<string, unknown>).responseMimeType).toBe('application/json');
    const contents = body.contents as Array<{ parts: Array<{ text: string }> }>;
    expect(contents[0]!.parts[0]!.text).toBe(buildReviewMessages(reviewPayload).userContent);
  });
});

describe('GeminiNativeProvider.ask', () => {
  it('omits responseMimeType, uses ASK temperature, wraps result as { output_text }', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, geminiTextResponse('It will swap SOL for USDC via Jupiter.'));
    const provider = new GeminiNativeProvider(config(), http);
    const result = await provider.ask({ question: 'what happens?' });

    expect((result.output_text as string).startsWith('It will swap')).toBe(true);
    const body = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    const generationConfig = body.generationConfig as Record<string, unknown>;
    expect('responseMimeType' in generationConfig).toBe(false);
    expect(generationConfig.temperature).toBe(0.3);
    expect(generationConfig.maxOutputTokens).toBe(800);
  });

  it('attaches google_search when ask payload requests research', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, geminiTextResponse('Current Helium cheapest plan is $15.'));
    const provider = new GeminiNativeProvider(config(), http);
    await provider.ask({
      question: 'what is the current Helium plan?',
      research: { needed: true, mode: 'auto_current_facts', currentDate: '2026-05-17T00:00:00.000Z', maxSearches: 3 },
    });

    const body = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    expect(body.tools).toEqual([{ google_search: {} }]);
    expect('responseMimeType' in (body.generationConfig as Record<string, unknown>)).toBe(false);
  });

  it('throws provider_invalid_response when the model returns blank text', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, geminiTextResponse(''));
    const provider = new GeminiNativeProvider(config(), http);

    let captured: unknown = null;
    try {
      await provider.ask({ question: 'what happens?' });
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    expect((captured as ProviderHttpError).code).toBe('provider_invalid_response');
  });
});

describe('GeminiNativeProvider error handling', () => {
  it('403 → ProviderHttpError(provider_auth)', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(403, '{"error":{"message":"key not allowed"}}');
    const provider = new GeminiNativeProvider(config(), http);

    let captured: unknown = null;
    try {
      await provider.generatePlan({ userPrompt: 'hi' });
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    expect((captured as ProviderHttpError).code).toBe('provider_auth');
  });

  it('malformed 200 body becomes provider_invalid_response', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, '{not even json');
    const provider = new GeminiNativeProvider(config(), http);

    let captured: unknown = null;
    try {
      await provider.generatePlan({ userPrompt: 'hi' });
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    expect((captured as ProviderHttpError).code).toBe('provider_invalid_response');
  });
});

File: apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/GeminiNativeProviderTest.kt
package com.agentic.wallet.agent.provider

import com.agentic.wallet.agent.prompts.DeviceAgentMessageAssembler
import com.agentic.wallet.agent.prompts.DeviceAgentSystemPrompts
import com.agentic.wallet.agent.runtime.RuntimeConfig
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.runBlocking
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Clock
import java.time.Instant
import java.time.ZoneOffset

class GeminiNativeProviderTest {
    private fun config(
        model: String = "gemini-2.5-pro",
        baseUrl: String = "https://generativelanguage.googleapis.com/v1beta/openai",
    ): RuntimeConfig = RuntimeConfig(
        provider = "gemini",
        apiFormat = "openai-compatible",
        model = model,
        baseUrl = baseUrl,
        apiKey = "AIzaTEST-ABCDEFGHIJKLMNOP",
        walletAddress = null,
    )

    private fun geminiTextResponse(text: String): String =
        """{"candidates":[{"content":{"parts":[{"text":"$text"}]}}]}"""

    @Test
    fun generatePlanPostsToGenerateContentWithSystemInstructionAndResponseSchema() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                geminiTextResponse("{\\\"intent\\\":\\\"swap\\\",\\\"route\\\":\\\"jupiter\\\",\\\"risk\\\":\\\"low\\\",\\\"approval\\\":\\\"once\\\",\\\"safeguards\\\":[\\\"check slippage\\\"]}"),
            )
        }
        val provider = GeminiNativeProvider(config(), http)
        val result = provider.generatePlan(JSONObject().put("userPrompt", "swap 1 SOL for USDC"))

        assertEquals("swap", result.optString("intent"))
        val call = http.calls.single()
        // /openai stripped from preset, model + :generateContent appended.
        assertEquals(
            "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-pro:generateContent",
            call.url,
        )
        assertEquals("AIzaTEST-ABCDEFGHIJKLMNOP", call.headers["x-goog-api-key"])
        assertNull(call.headers["Authorization"])

        val body = JSONObject(call.body)
        val systemInstruction = body.optJSONObject("systemInstruction")!!
        assertEquals(
            DeviceAgentSystemPrompts.PLAN,
            systemInstruction.optJSONArray("parts")?.optJSONObject(0)?.optString("text"),
        )
        val contents = body.optJSONArray("contents")!!
        assertEquals(1, contents.length())
        assertEquals("user", contents.getJSONObject(0).optString("role"))
        assertEquals(
            DeviceAgentMessageAssembler.buildPlanMessages(JSONObject().put("userPrompt", "swap 1 SOL for USDC")).userContent,
            contents.getJSONObject(0).optJSONArray("parts")?.optJSONObject(0)?.optString("text"),
        )

        val generationConfig = body.optJSONObject("generationConfig")!!
        assertEquals(0.2, generationConfig.optDouble("temperature"), 0.0001)
        assertEquals(1024, generationConfig.optInt("maxOutputTokens"))
        assertEquals("application/json", generationConfig.optString("responseMimeType"))
        val responseSchema = generationConfig.optJSONObject("responseSchema")!!
        assertEquals("object", responseSchema.optString("type"))
        assertTrue(responseSchema.optJSONArray("required")!!.toString().contains("safeguards"))
        assertFalse(body.has("tools"))
    }

    @Test
    fun generatePlanIsIdempotentWithNativeBaseUrl() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, geminiTextResponse("{\\\"intent\\\":\\\"x\\\"}"))
        }
        val provider = GeminiNativeProvider(
            config(baseUrl = "https://generativelanguage.googleapis.com/v1beta"),
            http,
        )
        provider.generatePlan(JSONObject().put("userPrompt", "hi"))
        assertEquals(
            "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-pro:generateContent",
            http.calls.single().url,
        )
    }

    @Test
    fun generatePlanAppendsV1betaWhenBareHost() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, geminiTextResponse("{\\\"intent\\\":\\\"x\\\"}"))
        }
        val provider = GeminiNativeProvider(
            config(model = "gemini-2.5-flash", baseUrl = "https://example-gemini-proxy.dev"),
            http,
        )
        provider.generatePlan(JSONObject().put("userPrompt", "hi"))
        assertEquals(
            "https://example-gemini-proxy.dev/v1beta/models/gemini-2.5-flash:generateContent",
            http.calls.single().url,
        )
    }

    @Test
    fun reviewPlanRunsTwoPassResearchWithGoogleSearchAndNoJsonResponseConfig() = runBlocking {
        val http = FakeHttpExecutor().apply {
            // Pass 1: grounded research returns text + groundingMetadata.
            queueResponse(
                200,
                """{"candidates":[{"content":{"parts":[{"text":"Helium Mobile cheapest plan is Air at $15/month."}]},"groundingMetadata":{"groundingChunks":[{"web":{"uri":"https://www.heliummobile.com/plans","title":"Plans — Helium Mobile"}}]}}]}""",
            )
            // Pass 2: structured review.
            queueResponse(
                200,
                geminiTextResponse("{\\\"decision\\\":\\\"approve\\\",\\\"reason\\\":\\\"Air plan is $15.\\\",\\\"summary\\\":\\\"Approved\\\",\\\"evidence\\\":{}}"),
            )
        }
        val fixedClock = Clock.fixed(Instant.parse("2026-05-18T00:00:00Z"), ZoneOffset.UTC)
        val provider = GeminiNativeProvider(config(), http, fixedClock)
        val result = provider.reviewPlan(
            JSONObject()
                .put("instruction", "check helium mobile. lowest monthly plan. if less than $20. approve.")
                .put("plan", JSONObject().put("intent", "swap"))
                .put(
                    "research",
                    JSONObject().put("needed", true).put("mode", "auto_current_facts"),
                ),
        )

        assertEquals("approve", result.optString("decision"))
        val evidence = result.optJSONObject("evidence")!!
        assertEquals("checked", evidence.optJSONObject("research")!!.optString("status"))
        assertTrue(evidence.optJSONArray("findings")!!.toString().contains("Current research"))
        assertTrue(evidence.optJSONArray("sources")!!.toString().contains("heliummobile.com"))
        assertEquals(2, http.calls.size)

        val researchBody = JSONObject(http.calls[0].body)
        val tools = researchBody.optJSONArray("tools")!!
        assertEquals(1, tools.length())
        assertTrue(tools.getJSONObject(0).has("google_search"))
        val researchGen = researchBody.optJSONObject("generationConfig")!!
        assertFalse(researchGen.has("responseMimeType"))
        assertFalse(researchGen.has("responseSchema"))
        assertEquals(1800, researchGen.optInt("maxOutputTokens"))

        val reviewBody = JSONObject(http.calls[1].body)
        assertFalse(reviewBody.has("tools"))
        val reviewGen = reviewBody.optJSONObject("generationConfig")!!
        assertEquals("application/json", reviewGen.optString("responseMimeType"))
        assertEquals(1800, reviewGen.optInt("maxOutputTokens"))
        val responseSchema = reviewGen.optJSONObject("responseSchema")!!
        assertEquals("object", responseSchema.optString("type"))
        assertTrue(responseSchema.optJSONObject("properties")!!.has("evidenceFactIds"))
        assertTrue(responseSchema.optJSONObject("properties")!!.has("blockingFactIds"))
        assertTrue(responseSchema.optJSONObject("properties")!!.has("missingFactIds"))
        assertTrue(responseSchema.optJSONObject("properties")!!.has("confidence"))
        val reviewInput =
            reviewBody.optJSONArray("contents")?.optJSONObject(0)?.optJSONArray("parts")?.optJSONObject(0)?.optString("text")
                .orEmpty()
        assertTrue(reviewInput.contains("researchEvidence"))
        assertTrue(reviewInput.contains("heliummobile.com"))
        assertTrue(
            "input should contain clock-fixed checkedAt; got: $reviewInput",
            reviewInput.contains("\"checkedAt\":\"2026-05-18T00:00:00Z\""),
        )
    }

    @Test
    fun reviewPlanFallsBackToSinglePassWhenResearchFails() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueFailure(RuntimeException("research pass network blip"))
            queueResponse(
                200,
                geminiTextResponse("{\\\"decision\\\":\\\"needs_input\\\",\\\"reason\\\":\\\"Cannot confirm.\\\",\\\"summary\\\":\\\"Needs input\\\",\\\"evidence\\\":{}}"),
            )
        }
        val provider = GeminiNativeProvider(config(), http)
        val result = provider.reviewPlan(
            JSONObject()
                .put("plan", JSONObject().put("intent", "swap"))
                .put(
                    "research",
                    JSONObject().put("needed", true).put("mode", "auto_current_facts"),
                ),
        )
        assertEquals("needs_input", result.optString("decision"))
        assertEquals(2, http.calls.size)
    }

    @Test
    fun singlePassReviewKeepsResponseMimeTypeAndOmitsTools() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                geminiTextResponse("{\\\"decision\\\":\\\"approve\\\",\\\"reason\\\":\\\"ok\\\",\\\"summary\\\":\\\"go\\\",\\\"evidence\\\":{}}"),
            )
        }
        val provider = GeminiNativeProvider(config(), http)
        val payload = JSONObject()
            .put("plan", JSONObject().put("intent", "swap"))
            .put(
                "research",
                JSONObject()
                    .put("needed", false)
                    .put("mode", "not_required")
                    .put("currentDate", "2026-05-17T00:00:00Z")
                    .put("maxSearches", 3),
            )
        provider.reviewPlan(payload)

        val body = JSONObject(http.calls.single().body)
        assertFalse(body.has("tools"))
        val generationConfig = body.optJSONObject("generationConfig")!!
        assertEquals("application/json", generationConfig.optString("responseMimeType"))
        assertTrue(generationConfig.optJSONObject("responseSchema")!!.optJSONObject("properties")!!.has("evidenceFactIds"))
        val input = body.optJSONArray("contents")?.optJSONObject(0)?.optJSONArray("parts")?.optJSONObject(0)?.optString("text")
        assertEquals(DeviceAgentMessageAssembler.buildReviewMessages(payload).userContent, input)
    }

    @Test
    fun askOmitsResponseMimeTypeAndUsesAskTemperature() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, geminiTextResponse("It will swap SOL for USDC via Jupiter."))
        }
        val provider = GeminiNativeProvider(config(), http)
        val result = provider.ask(JSONObject().put("question", "what happens?"))

        assertTrue(result.optString("output_text").startsWith("It will swap"))
        val gen = JSONObject(http.calls.single().body).optJSONObject("generationConfig")!!
        assertFalse(gen.has("responseMimeType"))
        assertFalse(gen.has("responseSchema"))
        assertEquals(0.3, gen.optDouble("temperature"), 0.0001)
        assertEquals(800, gen.optInt("maxOutputTokens"))
    }

    @Test
    fun askAttachesGoogleSearchWhenResearchRequested() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, geminiTextResponse("Current Helium cheapest plan is $15."))
        }
        val provider = GeminiNativeProvider(config(), http)
        provider.ask(
            JSONObject()
                .put("question", "what is the current Helium plan?")
                .put(
                    "research",
                    JSONObject().put("needed", true).put("mode", "auto_current_facts"),
                ),
        )

        val body = JSONObject(http.calls.single().body)
        val tools = body.optJSONArray("tools")!!
        assertTrue(tools.getJSONObject(0).has("google_search"))
        val generationConfig = body.optJSONObject("generationConfig")!!
        assertFalse(generationConfig.has("responseMimeType"))
        assertFalse(generationConfig.has("responseSchema"))
    }

    @Test
    fun askThrowsInvalidResponseOnBlankText() {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, geminiTextResponse(""))
        }
        val provider = GeminiNativeProvider(config(), http)
        val ex = try {
            runBlocking { provider.ask(JSONObject().put("question", "what happens?")) }
            null
        } catch (e: ProviderHttpException) {
            e
        }
        assertNotEquals(null, ex)
        assertEquals(ProviderErrorCodes.INVALID_RESPONSE, ex!!.code)
    }

    @Test
    fun cancellationDuringResearchPassPropagates() {
        val http = FakeHttpExecutor().apply {
            queueFailure(CancellationException("user backed out"))
        }
        val provider = GeminiNativeProvider(config(), http)
        val ex = runCatching {
            runBlocking {
                provider.reviewPlan(
                    JSONObject()
                        .put("plan", JSONObject().put("intent", "swap"))
                        .put(
                            "research",
                            JSONObject().put("needed", true).put("mode", "auto_current_facts"),
                        ),
                )
            }
        }.exceptionOrNull()
        assertTrue("Expected CancellationException, got $ex", ex is CancellationException)
        assertEquals(1, http.calls.size)
    }

    @Test
    fun authErrorMapsTo403Code() {
        val http = FakeHttpExecutor().apply {
            queueResponse(403, """{"error":{"message":"key not allowed"}}""")
        }
        val provider = GeminiNativeProvider(config(), http)
        val ex = try {
            runBlocking { provider.generatePlan(JSONObject().put("userPrompt", "hi")) }
            null
        } catch (e: ProviderHttpException) {
            e
        }
        assertNotEquals(null, ex)
        assertEquals(ProviderErrorCodes.AUTH, ex!!.code)
    }
}
